src/tridiag.jl

```diff
@@ -0,0 +1,458 @@
+
+
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+#### Specialized matrix types ####
+
+struct SkewHermTridiagonal{T, V<:AbstractVector{T}} <: AbstractMatrix{T}
+    ev::V                        # subdiagonal
+    function SkewHermTridiagonal{T, V}(ev) where {T, V<:AbstractVector{T}}
+        require_one_based_indexing(ev)
+        new{T, V}(ev)
+    end
+end
+
+"""
+    SkewHermTridiagonal(ev::V) where V <: AbstractVector
+Construct a skewhermitian tridiagonal matrix from the subdiagonal (`ev`).
+The result is of type `SkewHermTridiagonal`
+and provides efficient specialized eigensolvers, but may be converted into a
+regular matrix with [`convert(Array, _)`](@ref) (or `Array(_)` for short).
+# Examples
+```jldoctest
+julia> ev = [7, 8, 9]
+3-element Vector{Int64}:
+ 7
+ 8
+ 9
+julia> SkewHermTridiagonal(ev)
+4×4 SkewHermTridiagonal{Int64, Vector{Int64}}:
+ 0 -7  ⋅  ⋅
+ 7  . -8  ⋅
+ ⋅  8  . -9
+ ⋅  ⋅  9  .
+```
+"""
+SkewHermTridiagonal(ev::V) where {T,V<:AbstractVector{T}} = SkewHermTridiagonal{T}(ev)
+SkewHermTridiagonal{T}(ev::V) where {T,V<:AbstractVector{T}} = SkewHermTridiagonal{T,V}(ev)
+function SkewHermTridiagonal{T}(ev::AbstractVector) where {T}
+    SkewHermTridiagonal(convert(AbstractVector{T}, ev)::AbstractVector{T})
+end
+
+"""
+    SkewHermTridiagonal(A::AbstractMatrix)
+Construct a skewhermitian tridiagonal matrix from first subdiagonal
+of the skewhermitian matrix `A`.
+# Examples
+```jldoctest
+julia> A = [1 2 3; 2 4 5; 3 5 6]
+3×3 Matrix{Int64}:
+ 1  2  3
+ 2  4  5
+ 3  5  6
+julia> SkewHermTridiagonal(A)
+3×3 SkewHermTridiagonal{Int64, Vector{Int64}}:
+ . -2  ⋅
+ 2  . -5
+ ⋅  5  .
+```
+"""
+function SkewHermTridiagonal(A::AbstractMatrix)
+    if diag(A, 1) == - adjoint.(diag(A, -1))
```

Comment:
This allocates 4 arrays and then discards them; not sure if we care, but it's a bit excessive for a check.

You can remove 1 allocation by saving `ev = diag(A,1)` (since you re-use it on the same line), and another allocation by using `.-` instead of `-` (since `.-` will "fuse" with the `adjoint.(...)` call).